### ISSUE TITLE
Allow user-controlled private keys

### ISSUE THREAD

**Issue by jas4711** [tillitis/tkey-device-signer#9]
While tillitis/tkey-device-signer#8 would be nice for me, I realize some use-cases that requires me to control the private key generation myself, and would thus not even want to generate private keys on the device at all.   I couldn't find a feature request to support this, so I'm opening this for discussion.  I think the simplest approach is to resolve tillitis/tillitis-key1-apps#34 and then "import" private keys into the device by encrypting them once, and then later supply the encrypted blob which is decrypted by the device, and its private-key operation is performed, and the output returned.  Thoughts?

**Comment by solemnwarning** [comment]
Another vote for this - I think the TKey would be a good device for signing update feeds in a project I'm working on, however I would want to be able to generate (and back up) the key myself so that a broken TKey doesn't leave me unable to sign new updates.

**Comment by udf2457** [comment]
@SallSim Castor sounds exciting ! 

The present Tkey format is great, but the main point of friction for many applications from a purely practical perspective is that the app updates break functionality.   This puts Tkey at a disadvantage to other offerings on the market (e.g. Yubikey or Nitrokey).

Of course for technical users its no problem to migrate from older to newer versions of Tkey apps.  But if you are thinking about more widespread use to a non-tech userbase then the experience needs to be a bit more foolproof. 😉

**Comment by SallSim** [comment]
@udf2457 nice you like the Castor idea!

Yes, updating an app today changes identity and that can be a drawback. The basic idea is that you *can* update the app with e.g. a new authentication protocol, whereas with other vendors you'd need to buy new hardware key to make use new features.

With that said, we are looking into simplifying. We have started work [here](https://github.com/tillitis/tkey-boot-verifier).

**Comment by udf2457** [comment]
@SallSim True. But e.g. with Yubi you can generate a key on-token or load one.  So you can e.g. setup multiple Yubis with the same key (the YubiHSM provides wrapped copies, but for plain Yubi you need to load your own e.g. via an offline box ).

With Tkey at present this would not be achievable, even if the core app functionality (e.g. algorithm) didn't change.  Although perhaps it would be fundamentally unachievable due to the overall Tkey architecture design ?

P.S. Your `tilitis/tkey-boot-verifier` link gives a 404 ?
